In [5]:
from dotenv import load_dotenv
import os
import sys

sys.path.append('../')

load_dotenv()

config = {
    "base_url": os.getenv("BASE_URL"),
    "client_id": os.getenv("CLIENT_ID"),
    "client_secret": os.getenv("CLIENT_SECRET")
}


In [6]:
import re

def get_page(url):
    """
    Devuelve el número de página de una URL.
    """
    match = re.search(r'page=(\d+)', url)
    if match:
        return int(match.group(1))

url = 'https://sandbox.belvo.com/api/transactions/?link=e0f02811-39f3-4c08-be2d-b65476e13ff5&page=2'

page_number = get_page(url)
print(f"Página: {page_number}")

Página: 2


In [ ]:
from services.ofda_service import OFDAService

ofda_service = OFDAService(config)
all_transactions = []
link_id = 'e0f02811-39f3-4c08-be2d-b65476e13ff5'

page= 1

while page:
    transactions = ofda_service.get_transactions(link_id,page)
    all_transactions.extend(transactions['results'])
    page = get_page(transactions['next']) if transactions['next'] else None

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


In [10]:
len(all_transactions)

2294

In [12]:
import pandas as pd 
df = pd.DataFrame(all_transactions)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2294 entries, 0 to 2293
Data columns (total 29 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              2294 non-null   object 
 1   account                         2294 non-null   object 
 2   created_at                      2294 non-null   object 
 3   category                        0 non-null      object 
 4   subcategory                     0 non-null      object 
 5   merchant                        0 non-null      object 
 6   mcc                             0 non-null      object 
 7   type                            2294 non-null   object 
 8   amount                          2294 non-null   float64
 9   status                          2294 non-null   object 
 10  balance                         0 non-null      object 
 11  currency                        2294 non-null   object 
 12  loan_data                       0 

In [14]:
df = df.dropna(axis=1, how='any').copy()

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2294 entries, 0 to 2293
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        2294 non-null   object 
 1   account                   2294 non-null   object 
 2   created_at                2294 non-null   object 
 3   type                      2294 non-null   object 
 4   amount                    2294 non-null   float64
 5   status                    2294 non-null   object 
 6   currency                  2294 non-null   object 
 7   value_date                2294 non-null   object 
 8   description               2294 non-null   object 
 9   collected_at              2294 non-null   object 
 10  counterparty              2294 non-null   object 
 11  transacted_at             2294 non-null   object 
 12  operation_type            2294 non-null   object 
 13  internal_identification   2294 non-null   object 
 14  inferred

In [16]:
df.head()

id  \
0  00017ce9-4a85-49bf-9e21-2842f06ff5d5   
1  0029a175-8a99-4695-beb7-bfccf50365cf   
2  006f7aa4-c8b0-4416-aa80-8373badeb0e4   
3  0077a313-08e9-4258-ab1a-f9ebc330ac46   
4  0084da46-d916-429d-a33b-6726378e6156   

                                             account  \
0  {'id': '5b2073be-5ed5-48bb-8d42-242c9402eb8d',...   
1  {'id': '5b2073be-5ed5-48bb-8d42-242c9402eb8d',...   
2  {'id': '5b2073be-5ed5-48bb-8d42-242c9402eb8d',...   
3  {'id': '50ae652a-41b9-44d9-a5cb-7ed6203c733a',...   
4  {'id': '50ae652a-41b9-44d9-a5cb-7ed6203c733a',...   

                    created_at     type  amount     status currency  \
0  2025-07-31T11:59:14.680861Z  OUTFLOW  771.52  PROCESSED      BRL   
1  2025-07-31T11:59:14.702401Z  OUTFLOW  771.52  PROCESSED      BRL   
2  2025-07-31T11:59:14.670951Z  OUTFLOW  771.52  PROCESSED      BRL   
3  2025-07-31T11:59:14.691694Z  OUTFLOW  771.52  PROCESSED      BRL   
4  2025-07-31T11:59:14.579266Z  OUTFLOW  771.52  PROCESSED      BRL   

   value_date description                 collected_at  \
0  2025-01-25   PAGAMENTO  2025-07-31T11:58:56.032624Z   
1  2025-01-22   PAGAMENTO  2025-07-31T11:58:56.407697Z   
2  2025-01-27   PAGAMENTO  2025-07-31T11:58:55.920124Z   
3  2025-01-23   PAGAMENTO  2025-07-31T11:59:10.827392Z   
4  2025-07-23   PAGAMENTO  2025-07-31T11:59:10.734838Z   

                                        counterparty  \
0  {'name': None, 'type': 'INDIVIDUAL', 'agency':...   
1  {'name': None, 'type': 'INDIVIDUAL', 'agency':...   
2  {'name': None, 'type': 'INDIVIDUAL', 'agency':...   
3  {'name': None, 'type': 'INDIVIDUAL', 'agency':...   
4  {'name': None, 'type': 'INDIVIDUAL', 'agency':...   

              transacted_at operation_type  \
0  2025-01-25T03:01:13.613Z         CARTAO   
1  2025-01-22T03:01:13.613Z         CARTAO   
2  2025-01-27T03:01:13.613Z         CARTAO   
3  2025-01-23T03:01:13.613Z         CARTAO   
4  2025-07-23T03:01:13.613Z         CARTAO   

                internal_identification inferred_accounting_date  
0  20279336-7e3c-4628-8301-7f7ff43aa5a1               2025-01-27  
1  3b6839e5-1aeb-4f40-8faf-58b7729c3d56               2025-01-22  
2  2b686d54-8971-4df4-842a-c2bd0fb49f20               2025-01-27  
3  9096cb55-3b4f-4f7b-a1ff-a949bfe254ce               2025-01-23  
4  eb528819-f66b-44f5-9594-0496641374af               2025-07-23

In [17]:
df['type'].value_counts()

type
OUTFLOW    2294
Name: count, dtype: int64

In [18]:
df.status.value_counts()

status
PROCESSED    2294
Name: count, dtype: int64

In [19]:
df.currency.value_counts()

currency
BRL    2294
Name: count, dtype: int64

In [20]:
min(df.value_date), max(df.value_date)

('2024-08-05', '2025-07-28')

In [21]:
df.operation_type.value_counts()

operation_type
CARTAO    2294
Name: count, dtype: int64

In [22]:
df.counterparty.value_counts()

counterparty
{'name': None, 'type': 'INDIVIDUAL', 'agency': '6272', 'number': '94088392', 'check_digit': '4', 'clearing_code': '123', 'document_number': '87517400444'}    2294
Name: count, dtype: int64

In [23]:
df.account.value_counts()

account
{'id': '50ae652a-41b9-44d9-a5cb-7ed6203c733a', 'link': 'e0f02811-39f3-4c08-be2d-b65476e13ff5', 'institution': {'name': 'ofmockbank_br_retail', 'type': 'bank'}, 'created_at': '2025-07-31T11:35:36.384280Z', 'name': 'Sib Bank', 'type': 'CONTA_POUPANCA', 'agency': '6272', 'number': '11188222', 'balance': {'blocked': 999.99, 'current': 15000.0, 'available': 41233.07, 'automatically_invested': 15000.0}, 'subtype': 'INDIVIDUAL', 'category': 'SAVINGS_ACCOUNT', 'currency': 'BRL', 'overdraft': {'used': 12345.4, 'arranged': 100.0, 'unarranged': 100.0}, 'check_digit': '4', 'balance_type': 'ASSET', 'collected_at': '2025-07-31T12:45:59.879051Z', 'clearing_code': '123', 'internal_identification': '291e5a29-49ed-401f-a583-193caa7acddd', 'public_identification_name': 'AGENCY/NUMBER', 'public_identification_value': '6272/11188222'}                     1234
{'id': '5b2073be-5ed5-48bb-8d42-242c9402eb8d', 'link': 'e0f02811-39f3-4c08-be2d-b65476e13ff5', 'institution': {'name': 'ofmockbank_br_retail'

# Transaction Data Analysis Summary

## Dataset Overview
- **Total Records**: 2,294 transactions
- **Date Range**: August 5, 2024 to July 28, 2025
- **Currency**: BRL (Brazilian Real)

## Data Distribution

### Transaction Types
- **OUTFLOW**: 2,294 records
- All transactions appear to be outgoing payments

### Transaction Status
- **PROCESSED**: 2,294 records
- All transactions have been successfully processed

### Operation Type
- **CARTAO**: 2,294 records
- All transactions are card-based operations

### Counterparty Information
- Contains structured data with fields:
  - `name`: Individual names (some None values)
  - `type`: "INDIVIDUAL" 
  - `agency`: "6272"
  - `number`: Various account numbers (e.g., "94088392")
  - `check_digit`: "4"
  - `clearing_code`: Present in records

### Account Information
- Multiple account identifications present:
  - Agency/Number combinations like "6272/11188222" and "6272/94088392"
  - Different account holders with varying identification numbers

## Key Insights
- This appears to be a dataset of card-based outflow transactions
- All transactions are in Brazilian Real (BRL)
- Data spans approximately 11 months
- All transactions have been successfully processed
- Consistent agency code (6272) suggests transactions from the same financial institution
